pkg/services/ngalert/api/api_admin.go

```diff
@@ -0,0 +1,62 @@
+package api
+
+import (
+	"errors"
+	"net/http"
+
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	apimodels "github.com/grafana/grafana/pkg/services/ngalert/api/tooling/definitions"
+	ngmodels "github.com/grafana/grafana/pkg/services/ngalert/models"
+	"github.com/grafana/grafana/pkg/services/ngalert/store"
+)
+
+type AdminSrv struct {
+	store store.AdminConfigurationStore
+	log   log.Logger
+}
+
+func (srv AdminSrv) RouteGetAdminConfig(c *models.ReqContext) response.Response {
+	cfg, err := srv.store.GetAdminConfiguration(c.OrgId)
+	if err != nil {
+		if errors.Is(err, store.ErrNoAdminConfiguration) {
+			return ErrResp(http.StatusNotFound, err, "")
+		}
+
+		msg := "failed to fetch admin configuration from the database"
+		srv.log.Error(msg, "err", err)
+		return ErrResp(http.StatusInternalServerError, err, msg)
+	}
+
+	resp := apimodels.AdminConfiguration{
+		Alertmanagers: cfg.Alertmanagers,
+	}
+	return response.JSON(http.StatusOK, resp)
+}
+
+func (srv AdminSrv) RouteUpdateAdminConfig(c *models.ReqContext, body apimodels.AdminConfiguration) response.Response {
+	cfg := &ngmodels.AdminConfiguration{
+		Alertmanagers: body.Alertmanagers,
+		OrgID:         c.OrgId,
+	}
+
+	cmd := store.UpdateAdminConfigurationCmd{AdminConfiguration: cfg}
+	if err := srv.store.UpdateAdminConfiguration(cmd); err != nil {
+		msg := "failed to save the admin configuration to the database"
+		srv.log.Error(msg, "err", err)
+		return ErrResp(http.StatusBadRequest, err, msg)
```

Comment:
Could the error also be 5xx in some cases?

Reply:
Technically it would, but `xorm` the ORM we use does not give me a different error between a failure of insertion due to payload vs not being able to communicate with the database. Any thoughts @papagian ?

Reply:
Each driver exposes the error code. So we could introduce similar errors like [IsUniqueConstraintViolation](https://github.com/grafana/grafana/blob/2c810506a57c694e54203775867779935468014d/pkg/services/sqlstore/migrator/dialect.go#L53) in the dialect interface (with [sqlite](https://github.com/grafana/grafana/blob/2c810506a57c694e54203775867779935468014d/pkg/services/sqlstore/migrator/sqlite_dialect.go#L144), [mysql](https://github.com/grafana/grafana/blob/2c810506a57c694e54203775867779935468014d/pkg/services/sqlstore/migrator/mysql_dialect.go#L186) and [postgres](https://github.com/grafana/grafana/blob/2c810506a57c694e54203775867779935468014d/pkg/services/sqlstore/migrator/postgres_dialect.go#L199) implementations) and catch them here.

Reply:
I think I'm going to skip this one until I get to API tests in my next PR - we're up to 90 comments on this one and it's getting hard to keep track of what.